### PR TITLE
ZTS: Add timeout to cp_stress

### PR DESCRIPTION
**Pull request by tonyhutter**
### Motivation and Context
`cp_stress` ZTS test can timeout after 10min on underpowered test systems

### Description
cp_stress is getting killed on the new QEMU-based github runners we're developing.  The problem is that the QEMU-based testers are so under-powered that the test is taking longer than the 10min maximum that ZTS enforces.  Instead, enforce an inter-test-cp_stress timeout so the entire test doesn't get killed.

### How Has This Been Tested?
Ran on FreeBSD 13 and Ubuntu 24.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: tests/zfs-tests/tests/functional/cp_files/cp_stress.ksh [hunk 1]
```diff
@@ -65,8 +65,17 @@ else
...
 for i in $(seq 1 $RUNS) ; do
-	# Each run takes around 12 seconds.
-	log_must $STF_SUITE/tests/functional/cp_files/seekflood 2000 $CPUS
+	# We have a maximum of 10min per ZTS test, so each seekflood run must
+	# be a fraction of that.  We need to add this timeout in so underpowered
+	# test runners don't get killed by ZTS and counted as a failure.
+	TO=$((600 / $RUNS))
+	timeout $TO $STF_SUITE/tests/functional/cp_files/seekflood 2000 $CPUS
+	rc=$?
+	if [ $rc == 124 ] ; then
```
Comment (robn) [inline tests/zfs-tests/tests/functional/cp_files/cp_stress.ksh:74]: Nit, might want to mention that `124` has a specific meaning for `timeout`.

**Review by robn — APPROVED** [review]
Looks right for what it is, but I wonder: what kind of "underpowered" are we talking here? Because `seekflood` is more likely to hit things if there are more CPUs. So if there's only one or two here, is it even worth running this?

**Comment by tonyhutter** [comment]
> Looks right for what it is, but I wonder: what kind of "underpowered" are we talking here? Because seekflood is more likely to hit things if there are more CPUs. So if there's only one or two here, is it even worth running this?

A little background: I'm currently helping test https://github.com/openzfs/zfs/pull/15838.  The idea there is to convert our buildbot builders to github runners.  Or more specifically, run ZTS within VMs on the github runners (using the runner as a VM host).  We need to use VMs, since github runners don't natively support all the OSs we test on (like Fedora, Almalinux, FreeBSD, etc).  In order to speed up the testing and not run into the 4-hour testing timeout, we're chunking up the the tests into thirds and running it on three VMs running in parallel.  That unfortunately means each each VM is going over-provisioned and run slower.  That's fine for most ZTS tests since they aren't very CPU bound, but there are some outliers like `cp_stress` that do run too long and get killed.

So yes, we ideally would want faster runners with more CPUs since they're going to expose race conditions and locking bugs like this quicker.  But we also have to make due with what's available.

File: tests/zfs-tests/tests/functional/cp_files/cp_stress.ksh (hunk 1 above)
Comment (tonyhutter) [reply tests/zfs-tests/tests/functional/cp_files/cp_stress.ksh:74]: I added a comment in my latest push:
```
# If 'timeout' returns a 124 return code, then it timed out.
```

**Comment by robn** [comment]
Oh I'm sorry, my question wasn't clear (also it was more of a drive-by mumble than a really big-brained thought).

I more meant, if we know this test is running in an environment where it's unlikely to actually bump into the thing it's checking for (because it isn't parallel enough), should we disable the test in that environment. Like, should it SKIP if it has fewer than N cores or something?

I suppose even "unlikely" is better than "never" though!

**Comment by tonyhutter** [comment]
I think it's still a useful test to run no matter what the hardware is; this popped up today: https://github.com/openzfs/zfs/issues/16297

**Comment by robn** [comment]
Oof, hard to argue with results. Also yikes, I'll check it out! :sweat_smile:

**Review by mcmilk — APPROVED** [review]

**Comment by mcmilk** [comment]
I created another pull request for this problem here: #16369

This one just takes down the RUNS to the limit of the FreeBSD testings, and then the test work fine on Linux also.
The cp_stress test needs around 2 minutes on FreeBSD and Linux then...

**Comment by mcmilk** [comment]
@tonyhutter - can be closed now :+1:

**Comment by tonyhutter** [comment]
Closing in favor of https://github.com/openzfs/zfs/pull/16369

**Pull request closed without merging**